### PR TITLE
[BITAU-124] Refresh Item List When Foregrounding the App

### PR DESCRIPTION
**Pull request by brant-livefront**
## 🎟️ Tracking

[BITAU-124](https://livefront.atlassian.net/browse/BITAU-124)

## 📔 Objective

This PR adds the NotificationCenterService and subscribes to the publisher to refresh the ItemList when the user returns to the app. Previously, if you made changes in the Bitwarden PM app and then switched back to the Authenticator, it would not have picked up those changes immediately. By subscribing to the foreground notification and performing a `.refresh` action, the user's latest changes from the PM app are immediately processed and shown.

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation team

## 🦮 Reviewer guidelines

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

### CODE REVIEW

File: AuthenticatorShared/UI/Vault/ItemList/ItemList/ItemListProcessorTests.swift
Comment (victor-livefront) [inline AuthenticatorShared/UI/Vault/ItemList/ItemList/ItemListProcessorTests.swift:70]: 👍

File: AuthenticatorShared/Core/Platform/Services/NotificationCenterService.swift
Comment (victor-livefront) [inline AuthenticatorShared/Core/Platform/Services/NotificationCenterService.swift:11]: Should we remove this definition since it's not being used?

File: AuthenticatorShared/Core/Platform/Services/NotificationCenterService.swift
Comment (victor-livefront) [inline AuthenticatorShared/Core/Platform/Services/NotificationCenterService.swift:34]: you could take this a step further and inject the NotificationCenter.default. Then have very basic tests to just make sure the MockNotificationCenter publishes through this publisher.

File: AuthenticatorShared/Core/Platform/Services/NotificationCenterService.swift
Comment (brant-livefront) [reply AuthenticatorShared/Core/Platform/Services/NotificationCenterService.swift:11]: I debated that as well. I felt like keeping it since they are kind of a pair and it might be useful in the future. But I'm also fine removing it since it's a fairly obvious thing to add when we do need it.

Comment (victor-livefront) [reply AuthenticatorShared/Core/Platform/Services/NotificationCenterService.swift:11]: I'd lean towards removing it because it's easy to add back, and may never be used.

**Comment by KatherineInCode** [comment]
⛏️ Could the PR title be changed to fix the typo?

File: AuthenticatorShared/Core/Platform/Services/NotificationCenterService.swift
Comment (KatherineInCode) [reply AuthenticatorShared/Core/Platform/Services/NotificationCenterService.swift:11]: I'd vaguely lean towards keeping it to maintain parity with the PM app

File: AuthenticatorShared/Core/Platform/Services/NotificationCenterService.swift
Comment (KatherineInCode) [reply AuthenticatorShared/Core/Platform/Services/NotificationCenterService.swift:34]: It would be good to backport those tests/implementation to the PM app at some point, if we do.

File: AuthenticatorShared/UI/Vault/ItemList/ItemList/ItemListProcessor.swift
Comment (KatherineInCode) [inline AuthenticatorShared/UI/Vault/ItemList/ItemList/ItemListProcessor.swift:273]: 🤔 Why not a `for await _ in services.notificationCenterService.willEnterForegroundPublisher() {` like we do in the PM app?

File: AuthenticatorShared/Core/Platform/Services/NotificationCenterService.swift
Comment (victor-livefront) [reply AuthenticatorShared/Core/Platform/Services/NotificationCenterService.swift:11]: Good point. Works for me.

File: AuthenticatorShared/UI/Vault/ItemList/ItemList/ItemListProcessor.swift
Comment (brant-livefront) [reply AuthenticatorShared/UI/Vault/ItemList/ItemList/ItemListProcessor.swift:273]: That's what I had originally, but I was running into lots of threading issues especially with the tests. We already have one `for await` (in streamItemList) that never terminates, so calling them both on `.appear` won't work. Calling it in a `Task` in `init` was causing the refresh to occur in other tests and other big threading problems. So it felt simpler to just use it directly as `AnyPublisher` rather than trying to map it into `for await`. Using the non-async version guarantees that we get the notification subscription set up (because it's synchronous) and we only see the single refresh that we expect in the tests.

File: AuthenticatorShared/Core/Platform/Services/NotificationCenterService.swift
Comment (brant-livefront) [reply AuthenticatorShared/Core/Platform/Services/NotificationCenterService.swift:51]: Good suggestion. I added a commit with tests to cover this and I have a similar set of tests in a branch for the PM app that I'll put up after this is approved. 👍

File: AuthenticatorShared/UI/Vault/ItemList/ItemList/ItemListProcessor.swift
Comment (KatherineInCode) [reply AuthenticatorShared/UI/Vault/ItemList/ItemList/ItemListProcessor.swift:273]: Interesting. Alright, makes sense.

File: AuthenticatorShared/UI/Vault/ItemList/ItemList/ItemListProcessor.swift
Comment (fedemkr) [inline AuthenticatorShared/UI/Vault/ItemList/ItemList/ItemListProcessor.swift:279]: ⛏️ I think we should just break sooner if `self` is nil with a guard so it doesn't create the `Task` in that case.
```suggestion
            .sink { [weak self] in
                guard let self else { return }
                Task {
                    await self.perform(.refresh)
                }
            }
```

**Review by fedemkr — APPROVED** [review]
LGTM, just one minor pick but leave it up to you whether to change it.

**Review by fedemkr — DISMISSED** [review]

**Pull request merged**
